lib/internal/messages.js

```diff
@@ -0,0 +1,155 @@
+'use strict';
+
+/**
+ * Provides a simple internal API useful for externalizing Node.js' own
+ * output. The messages are defined as string templates compiled into
+ * the Node binary itself. The format is straightforward:
+ *
+ * const I18N = require('internal/messages');
+ *
+ * Message keys are integers defined as constants on the I18N object:
+ *   e.g. I18N.ERR_FOO, I18N.ERR_BAR, etc
+ *
+ * Given a hypothetical key ERR_FOO, with a message "Hello {0}",
+ *
+ *   const I18N = require('internal/messages');
+ *   console.log(I18N(I18N.ERR_FOO, 'World')); // Prints: Hello World
+ *
+ * If a particular message is expected to be used multiple times, the
+ * special `prepare()` method can be used to create a shortcut function:
+ *
+ *   const I18N = require('internal/messages');
+ *   const hello = I18N.prepare(I18N.ERR_FOO);
+ *   console.log(hello('World')); // Prints: Hello World
+ *
+ * As a convenience, it is possible to generate Error objects:
+ *
+ *    const I18N = require('internal/messages');
+ *    throw I18N.TypeError('I18N.ERR_FOO, 'World');
+ *    throw I18N.Error(I18N.ERR_FOO, 'World');
+ *    throw I18N.RangeError(I18N.ERR_FOO, 'World');
+ *
+ * These Error objects will include both the expanded message and the
+ * string version of the message key in the Error.message, and will
+ * include a `.key` property whose value is the string version of the
+ * message key.
+ *
+ * Calls to console.log() and console.error() can also be replaced with
+ * externalized versions:
+ *
+ *    const I18N = require('internal/messages');
+ *    I18N.log(I18N.ERR_FOO, 'World');
+ *      // Prints: `Hello World (ERR_FOO)` to stdout
+ *    I18N.error(I18N.ERR_FOO, 'World');
+ *      // Prints: `Hello World (ERR_FOO)` to stderr
+ *
+ **/
+
+const messages = process.binding('messages');
+
+function replacement(msg, values) {
+  if (values.length > 0)
+    msg = msg.replace(/{(\d)}/g, (_, index) => values[Number(index)]);
```

Comment:
Meaning for the messages in general? I originally went with that approach but it ends up being a bit problematic in that it uses the actual message text as the look up key, which means (a) it's less efficient on the lookup and (b) it adds more work to keep things in sync. This approach ends up being more efficient overall.